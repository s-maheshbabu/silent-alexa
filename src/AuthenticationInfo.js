import { hasIn } from "immutable";
import util from "util";

export default class AuthenticationInfo {
  constructor(authResponse) {
    if (arguments.length == 1 && this.validateAuthResponse(authResponse)) {
      var _access_token = authResponse.access_token;
      var _expires_in = authResponse.expires_in;
    }
    this.getAccessToken = function() {
      return _access_token;
    };
  }

  isValid() {
    return this.getAccessToken() ? true : false;
  }

  validateAuthResponse(authResponse) {
    if (
      hasIn(authResponse, ["access_token"]) &&
      hasIn(authResponse, ["expires_in"]) &&
      authResponse.access_token &&
      authResponse.expires_in
    ) {
      return true;
    } else {
      console.log(
        "Encountered an error on login: " +
          util.inspect(authResponse, { showHidden: true, depth: null })
      );
      return false;
    }
  }
}
