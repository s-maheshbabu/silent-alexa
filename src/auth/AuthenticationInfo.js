import IllegalArgumentError from "errors/IllegalArgumentError";
import Cookies from "js-cookie";
import util from "util";

/**
 * Wrapper class to hold access_token and expires_in returned in LoginWithAmazon
 * authorization response.
 * Instances are created only if access_token and expires_in are defined. The token
 * is stored in a cookie called 'amazon_Login_accessToken' because that is the
 * cookie that LoginWithAmazon will look for.
 * https://developer.amazon.com/docs/login-with-amazon/javascript-sdk-reference.html#setUseCookie
 */
// TODO: Rename this to AuthenticationManager now that the class has been refactored to not
// actually hold the authentication information.
class AuthenticationInfo {
  persist(lwaResponse) {
    const numberOfSecondsInADay = 86400;

    if (lwaResponse && lwaResponse.access_token && lwaResponse.expires_in) {
      Cookies.set(AMAZON_LOGIN_COOKIE, lwaResponse.access_token, {
        expires: lwaResponse.expires_in / numberOfSecondsInADay,
        secure: false // TODO: Change localhost to also use https and then change this to true.
      });
    } else {
      const serializedLWAResponse = util.inspect(lwaResponse, {
        showHidden: true,
        depth: null
      });
      throw new IllegalArgumentError(
        `LoginWithAmazon Authorization response is undefined or 
          doesnt have access_token/expires_in.
          lwaResponse: ${serializedLWAResponse}`
      );
    }
  }

  /**
   * @returns true is authentication info is present and false otherwise.
   * This does not guarantee that the tokens haven't already expired.
   * Callers should be able to handle the case of tokens being
   * invalid when they try to use them.
   */
  isPresent = () => {
    return this.getAccessToken() !== undefined;
  };

  /**
   * @returns access_token to access Alexa API
   */
  getAccessToken = () => {
    return undefined;
  };

  /**
   * @returns deleted the cookie that stores the authentication info
   */
  clear = () => {
    Cookies.remove(AMAZON_LOGIN_COOKIE);
  };
}

// Declare a singleton
const instance = new AuthenticationInfo();
Object.freeze(instance);

export default instance;
export const AMAZON_LOGIN_COOKIE = "amazon_Login_accessToken";
