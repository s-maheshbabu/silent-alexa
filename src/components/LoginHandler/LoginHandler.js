import React from "react";
import queryString from "query-string";
import util from "util";
import AuthenticationInfo from "AuthenticationInfo";
import { hasIn } from "immutable";

export default class LoginHandler extends React.Component {
  render() {
    return null;
  }

  componentWillMount() {
    let lwaResponse;
    // Parses the query string to fetch the login with amazon response object
    if (hasIn(this.props, ["location", "hash"])) {
      lwaResponse = queryString.parse(this.props.location.hash);
    }
    if (this.isLWAResponseValid(lwaResponse)) {
      this.props.updateAuthenticationInfo(new AuthenticationInfo(lwaResponse));
    }

    if (this.props.history) {
      this.props.history.push("/");
    }
  }

  /**
   * @returns true if LoginWithAmazon response is valid
   *          false, otherwise.
   */
  isLWAResponseValid(lwaResponse) {
    if (
      !lwaResponse ||
      !lwaResponse.access_token ||
      !lwaResponse.expires_in ||
      lwaResponse.expires_in < 0
    ) {
      console.log(
        "Encountered an error on login: " +
          util.inspect(lwaResponse, { showHidden: true, depth: null })
      );
      return false;
    }
    return true;
  }
}
