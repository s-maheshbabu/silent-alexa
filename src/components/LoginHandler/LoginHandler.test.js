import React from "react";
import { shallow, mount } from "enzyme";
import queryString from "query-string";
import LoginHandler from "./LoginHandler";
import AuthenticationInfo from "AuthenticationInfo";

jest.mock("AuthenticationInfo");

beforeEach(() => {
  jest.resetAllMocks();
});

it("renders LoginHandler without crashing", () => {
  const wrapper = mount(<LoginHandler />);

  expect(wrapper).toMatchSnapshot();

  wrapper.unmount();
});

it("expects AuthenticationInfo to be persisted when LoginHandler is mounted", () => {
  const lwaResponseHash = "access_token=some_access_token&expires_in=30";
  const routeProps = { location: { hash: lwaResponseHash } };
  const wrapper = mount(<LoginHandler {...routeProps} />);

  expect(AuthenticationInfo.persist).toHaveBeenCalledTimes(1);
  expect(AuthenticationInfo.persist).toHaveBeenCalledWith(
    queryString.parse(lwaResponseHash)
  );

  wrapper.unmount();
});

it("expects AuthenticationInfo persist to not be called when LoginHandler is mounted with invalid parameters", () => {
  const lwaResponseWithNoAccessToken =
    "error_token=some_access_token&expires_in=30";
  const lwaResponseWithNoExpiresIn = "access_token=some_access_token";
  const lwaResponseWithInvalidExpiresIn =
    "access_token=some_access_token&expires_in=-30";
  const lwaResponse = "access_token=some_access_token&expires_in=30";
  const routePropsWithlwaResponseNoAccessToken = {
    location: { hash: lwaResponseWithNoAccessToken }
  };
  const routePropsWithlwaResponseNoExpiresIn = {
    location: { hash: lwaResponseWithNoExpiresIn }
  };
  const routePropsWithlwaResponseInvalidExpiresIn = {
    location: { hash: lwaResponseWithInvalidExpiresIn }
  };
  const routePropsNoLocation = { not_location: { hash: lwaResponse } };
  const routePropsNoHash = { location: { not_hash: lwaResponse } };
  const routePropsNoLwaResponse = { location: { hash: {} } };

  const allRoutePropsObjects = [
    routePropsWithlwaResponseNoAccessToken,
    routePropsWithlwaResponseNoExpiresIn,
    routePropsWithlwaResponseInvalidExpiresIn,
    routePropsNoLocation,
    routePropsNoHash,
    routePropsNoLwaResponse
  ];

  allRoutePropsObjects.map(routeProps => {
    const wrapper = mount(<LoginHandler {...routeProps} />);
    expect(AuthenticationInfo.persist).not.toHaveBeenCalled();
    wrapper.unmount();
  });
});

it("expects AuthenticationInfo persist to not be called when LoginHandler is mounted without route props", () => {
  const wrapperWithNoRouteProps = mount(<LoginHandler />);
  expect(AuthenticationInfo.persist).not.toHaveBeenCalled();
  wrapperWithNoRouteProps.unmount();
});
