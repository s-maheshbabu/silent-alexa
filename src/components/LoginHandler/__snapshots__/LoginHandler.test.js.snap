// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders LoginHandler without crashing 1`] = `<LoginHandler />`;
